[FEAT] Implement journey planner feature with API for testing purposes

ninouGx opened the pull request.

ninouGx committed:
[25] (feat on coord-api) Implement journey planner feature with API integration and UI components - testing purposes

TomPlanche commented on src/routes/api/journey/+server.ts:
```diff
@@ -0,0 +1,69 @@
+import type { RequestHandler } from "./$types";
+import axios from "axios";
+import { z } from "zod";
+import { PUBLIC_BACK_ENDPOINT } from "$env/static/public";
```
Put it in the `env.example` too.

TomPlanche commented on src/lib/schemas/journey.ts:
```diff
@@ -0,0 +1,37 @@
+});
+
+const SectionSchema = z.object({
+  duration: z.number(),
```
This allows negative duration, use `duration: z.number().positive("Duration must be positive")`

TomPlanche commented on src/lib/schemas/journey.ts:
```diff
@@ -0,0 +1,37 @@
+  arrival_date_time: z.string(),
+  from: PlaceSchema,
+  to: PlaceSchema,
+  type: z.string(),
```
maybe use an enum here

TomPlanche commented on src/lib/schemas/journey.ts:
```diff
@@ -0,0 +1,37 @@
+import { z } from "zod";
+
+const CoordinatesSchema = z.object({
+  lat: z.string(),
```
this allows any string, add a `.refine` to check if it's a valid float and a valid coord (between -90 and 90 for lat, -180 and 180 for lon)

TomPlanche commented on src/lib/schemas/journey.ts:
Maybe have a string validator for dates, something like:
```
/**
 * Validates ISO 8601 date strings
 */
const dateTimeSchema = z.string().refine(
  (val) => !isNaN(Date.parse(val)),
  "Invalid date format. Must be ISO 8601"
);
```

TomPlanche commented on src/lib/schemas/journey.ts:
```diff
@@ -0,0 +1,37 @@
+});
+
+const JourneySchema = z.object({
+  duration: z.number(),
```
Same remark as for `SectionSchema`

TomPlanche commented on src/lib/services/journeyServices.ts:
```diff
@@ -0,0 +1,70 @@
+    toLat: number
+  ): Promise<TJourneysResponse> {
+    try {
+      const fromCoords = `${fromLon};${fromLat}`;
```
Maybe have a utility function like:
```
const formatCoordinates = (coords: Coordinates): string => {
    return `${coords.lon};${coords.lat}`;
  }
```

With a type:
```
type Coordinates = {
  lon: number;
  lat: number;
};
```

TomPlanche commented on src/lib/services/journeyServices.ts:
```diff
@@ -0,0 +1,70 @@
+      // Validate response against schema
+      return JourneysResponseSchema.parse(response.data.data);
+    } catch (e) {
+      if (e instanceof Error) {
```
Maybe have a dedicated func to handle this since it'll be shared behavior

TomPlanche commented on src/lib/stores/journey.ts:
Looks promising, maybe a refactor in the return section would be great after.

TomPlanche commented on src/routes/journey/+page.ts:
What's the purpose of this file ?

ninouGx replied:
It'll help to pre-load data later in the app that should be provide to the svelte page.

TomPlanche replied:
I think the translations are loaded by the 'src/routes/+layout.ts'

ninouGx replied:
Yeah I update my comment, it's for later usages, if we want to call the location of the user for example to pre-fill some fields for examples.

TomPlanche commented on src/routes/journey/+page.ts:
```diff
@@ -0,0 +1,6 @@
+import type { PageLoad } from "./$types";
+import { t } from "$lib/translations";
```
unused import

ninouGx committed:
(feat on coords-api) Enhance journey schema validation and formatting functions

- Improved latitude and longitude validation in CoordinatesSchema.
- Added utility functions for parsing and formatting Navitia date formats.
- Updated JourneyServices to use the new formatCoordinates function.
- Refactored date formatting in the journey page component.
- Added package manager specification in package.json.
- Updated example environment file with a backend endpoint.
